python/dgl/convert.py

```diff
@@ -429,94 +244,236 @@ def hetero_from_shared_memory(name):
+            utils.detect_inf_in_iterable(data[1], 'the destination node IDs for {}'.format(key))
+        else:
+            raise DGLError('Expect sequences (e.g., list, numpy.ndarray) or tensors for '
+                           'data_dict[{}], got {}'.format(key, type(data[0])))
+
+    data_types = set(data_types)
+    if len(data_types) != 1:
+        raise DGLError('Expect the node IDs to have a same type, got {}'.format(data_types))
+    # Skip the check for sequences
+    if len(data_dtypes) > 0:
+        data_dtypes = set(data_dtypes)
+        if len(data_dtypes) != 1:
+            raise DGLError('Expect the node tensors to have a same data type, '
+                           'got {}'.format(data_dtypes))
+        data_devices = set(data_devices)
+        if len(data_devices) != 1:
+            raise DGLError('Expect the node tensors to be on a same device, '
+                           'got {}'.format(data_devices))
+
+    utils.check_valid_idtype(idtype)
     if idtype is None:
-        for data in data_dict.values():
-            if isinstance(data, tuple) and len(data) == 2 and F.is_tensor(data[0]):
-                idtype = F.dtype(data[0])
-                break
+        if len(data_dtypes) > 0:
+            idtype = list(data_dtypes)[0]
+        else:
+            idtype = F.int64
 
-    # Convert all data to edge tensors first.
+    # Convert all data to node tensors first
     data_dict = {(sty, ety, dty) : utils.graphdata2tensors(data, idtype, bipartite=(sty != dty))
                  for (sty, ety, dty), data in data_dict.items()}
 
-    # infer number of nodes for each node type
+    # Sanity check for the node IDs and the number of nodes
+    num_nodes_dict_ = defaultdict(int)
+    for (srctype, etype, dsttype), data in data_dict.items():
+        relation = (srctype, etype, dsttype)
+        src, dst, nsrc, ndst = data
+        if len(src) > 0:
+            utils.assert_nonnegative_iterable(
+                src, 'the source node IDs of canonical edge type {}'.format(relation))
+            utils.assert_nonnegative_iterable(
+                dst, 'the destination node IDs of canonical edge type {}'.format(relation))
+        num_nodes_dict_[srctype] = max(num_nodes_dict_[srctype], nsrc)
+        num_nodes_dict_[dsttype] = max(num_nodes_dict_[dsttype], ndst)
+
     if num_nodes_dict is None:
-        num_nodes_dict = defaultdict(int)
-        for (srctype, etype, dsttype), data in data_dict.items():
-            _, _, nsrc, ndst = data
-            num_nodes_dict[srctype] = max(num_nodes_dict[srctype], nsrc)
-            num_nodes_dict[dsttype] = max(num_nodes_dict[dsttype], ndst)
+        num_nodes_dict = num_nodes_dict_
+    else:
+        utils.check_type(num_nodes_dict, dict, 'num_nodes_dict', skip_none=False)
+        for nty in num_nodes_dict_:
+            if nty not in num_nodes_dict:
+                raise DGLError('Missing node type {} for num_nodes_dict'.format(nty))
```

Comment:
If missing can we allow num_nodes_dict[nty] = num_nodes_dict_[nty].
It can be a case when user only what to specify the number of nodes for a specific node subset.

Reply:
Done.